org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/InferenceContext18.java

```diff
@@ -337,18 +337,18 @@ private void addInferenceVariables(InferenceVariable[] newVariables) {
 		}
 	}
 
-	/** Add new inference variables for the given type variables. */
-	public InferenceVariable[] addTypeVariableSubstitutions(TypeBinding[] typeVariables) {
-		int len2 = typeVariables.length;
+	/** Add new inference variables for the given type arguments. */
+	public InferenceVariable[] addTypeVariableSubstitutions(TypeBinding[] typeArguments) {
+		int len2 = typeArguments.length;
 		InferenceVariable[] newVariables = new InferenceVariable[len2];
 		InferenceVariable[] toAdd = new InferenceVariable[len2];
 		int numToAdd = 0;
-		for (int i = 0; i < typeVariables.length; i++) {
-			if (typeVariables[i] instanceof InferenceVariable)
-				newVariables[i] = (InferenceVariable) typeVariables[i]; // prevent double substitution of an already-substituted inferenceVariable
-			else
+		for (int i = 0; i < typeArguments.length; i++) {
+			if (typeArguments[i] instanceof InferenceVariable)
+				newVariables[i] = (InferenceVariable) typeArguments[i]; // prevent double substitution of an already-substituted inferenceVariable
+			else if (typeArguments[i] instanceof CaptureBinding)
```

Comment:
refresh from the spec: _the type variables produced by capture that are type arguments in Tc_

Reply:
will take a look at the more recent code that has
`			else if (!capturesOnly || typeArguments[i] instanceof CaptureBinding)
`